OrbitQt/TutorialOverlay.cpp

```diff
@@ -0,0 +1,429 @@
+      if (child != step.cutout_widget && child->parent() == tab) {
+        step.hints.push_back(DeriveHintDescription(cutout_rect, child));
+      }
+    }
+  }
+
+  return step;
+}
+
+void TutorialOverlay::InitAllStepsFromUi() {
+  steps_.clear();
+  for (int i = 0; i < ui_->tabWidget->count(); ++i) {
+    const std::string tab_name = ui_->tabWidget->widget(i)->objectName().toStdString();
+    CHECK(steps_.find(tab_name) == steps_.end());
+    steps_[tab_name] = InitializeStep(i);
+  }
+}
+
+const QRect TutorialOverlay::AbsoluteGeometry(QWidget* widget) const {
+  if (widget->parentWidget() == nullptr || widget->parentWidget() == parentWidget()) {
+    return widget->geometry();
+  }
+
+  QWidget* parent_widget = widget->parentWidget();
+  QRect result = widget->geometry();
+  QRect parent_geo = AbsoluteGeometry(parent_widget);
+
+  auto width = result.width();
+  auto height = result.height();
+
+  result.setX(result.x() + parent_geo.x());
+  result.setY(result.y() + parent_geo.y());
+
+  result.setWidth(width);
+  result.setHeight(height);
+
+  return result;
+}
+
+TutorialOverlay::~TutorialOverlay() {}
+
+void TutorialOverlay::StartActiveStep() {
+  const Step& step = GetActiveStep();
+  ui_->tabWidget->setCurrentIndex(step.tab_index);
+
+  if (step.setup.callback_init) {
+    step.setup.callback_init(this, step.name);
+  }
+
+  ui_->btnPrev->setEnabled(!ActiveStepIsFirstInSection());
+  const Section& section = GetActiveSection();
+  const std::string text = absl::StrFormat(
+      "%s: %d/%d", section.title, section.active_step_index + 1, section.step_names.size());
+  ui_->labelStep->setText(text.c_str());
+
+  if (step.setup.anchor_widget) {
+    step.setup.anchor_widget->installEventFilter(this);
+  }
+  showMaximized();
+  UpdateOverlayLayout();
+}
+
+bool TutorialOverlay::EndActiveStep() {
+  if (!HasActiveStep()) {
+    return true;
+  }
+
+  UnhookActiveStep();
+
+  return true;
+}
+
+void TutorialOverlay::StartSection(const std::string& section) {
+  if (!EndActiveStep()) {
+    return;
+  }
+
+  auto it = sections_.find(section);
+  CHECK(it != sections_.end());
+
+  active_section_ = it;
+  it->second.active_step_index = it->second.step_names.empty() ? -1 : 0;
+  StartActiveStep();
+}
+
+void TutorialOverlay::AddSection(std::string section_name, std::string title,
+                                 std::vector<std::string> step_names) {
+  Section section;
+  for (auto& name : step_names) {
+    CHECK(StepExists(name));
+  }
+  section.step_names = std::move(step_names);
+  section.title = std::move(title);
+  sections_[std::move(section_name)] = std::move(section);
+}
+
+void TutorialOverlay::NextStep() {
+  if (!HasActiveStep()) {
+    return;
+  }
+  if (!VerifyActiveStep()) {
+    return;
+  }
+  if (!EndActiveStep()) {
+    return;
+  }
+
+  auto& section = GetActiveSection();
+  section.active_step_index++;
+  if (section.active_step_index > section.step_names.size()) {
+    section.active_step_index = -1;
+  }
```

Comment:
Shouldn't we check for "section.active_step_index >= section.step_names.size()" ?

Reply:
Right, thanks. In fact it didn't make a difference because `HasActiveStep()` returns false when `active_step_index >= step_names.size()`, but still this isn't what I meant to achieve here.